test/e2e/query_test.go

```diff
@@ -120,9 +126,146 @@ func TestQuery(t *testing.T) {
+	ContainerSharedDir     = "/shared"
+)
+
+func createSDFile(sharedDir string, name string, fileSDStoreAddresses []string) (string, error) {
+	if len(fileSDStoreAddresses) > 0 {
+		queryFileSDDir := filepath.Join(sharedDir, "data", "querier", name)
+		container := filepath.Join(ContainerSharedDir, "data", "querier", name)
+		if err := os.MkdirAll(queryFileSDDir, 0750); err != nil {
+			return "", errors.Wrap(err, "create query dir failed")
+		}
+
+		fileSD := []*targetgroup.Group{{}}
+		for _, a := range fileSDStoreAddresses {
+			fileSD[0].Targets = append(fileSD[0].Targets, model.LabelSet{model.AddressLabel: model.LabelValue(a)})
+		}
+
+		b, err := yaml.Marshal(fileSD)
+		if err != nil {
+			return "", err
+		}
+
+		if err := ioutil.WriteFile(queryFileSDDir+"/filesd.yaml", b, 0600); err != nil {
+			return "", errors.Wrap(err, "creating query SD config failed")
+		}
+		return filepath.Join(container, "filesd.yaml"), nil
+	}
+	return "", nil
+}
+
+func TestQueryWithEndpointConfig(t *testing.T) {
+	t.Parallel()
+
+	e, err := e2e.NewDockerEnvironment("e2e_config")
+	testutil.Ok(t, err)
+	t.Cleanup(e2ethanos.CleanScenario(t, e))
+
+	receiver := e2ethanos.NewReceiveBuilder(e, "1").WithIngestionEnabled().Init()
+	testutil.Ok(t, e2e.StartAndWaitReady(receiver))
+
+	prom1, sidecar1 := e2ethanos.NewPrometheusWithSidecar(e, "alone", e2ethanos.DefaultPromConfig("prom-alone", 0, "", "", e2ethanos.LocalPrometheusTarget), "", e2ethanos.DefaultPrometheusImage(), "")
+	prom2, sidecar2 := e2ethanos.NewPrometheusWithSidecar(e, "remote-and-sidecar", e2ethanos.DefaultPromConfig("prom-both-remote-write-and-sidecar", 1234, e2ethanos.RemoteWriteEndpoint(receiver.InternalEndpoint("remote-write")), "", e2ethanos.LocalPrometheusTarget), "", e2ethanos.DefaultPrometheusImage(), "")
+	prom3, sidecar3 := e2ethanos.NewPrometheusWithSidecar(e, "ha1", e2ethanos.DefaultPromConfig("prom-ha", 0, "", "", e2ethanos.LocalPrometheusTarget), "", e2ethanos.DefaultPrometheusImage(), "")
+	prom4, sidecar4 := e2ethanos.NewPrometheusWithSidecar(e, "ha2", e2ethanos.DefaultPromConfig("prom-ha", 1, "", "", e2ethanos.LocalPrometheusTarget), "", e2ethanos.DefaultPrometheusImage(), "")
+	testutil.Ok(t, e2e.StartAndWaitReady(prom1, sidecar1, prom2, sidecar2, prom3, sidecar3, prom4, sidecar4))
+
+	// // Querier. Both fileSD and directly by flags.
+	fileSDPath, err := createSDFile(e.SharedDir(), "1", []string{sidecar3.InternalEndpoint("grpc"), sidecar4.InternalEndpoint("grpc")})
+	testutil.Ok(t, err)
+
+	endpointConfig := []query.EndpointConfig{
+		{
+			Endpoints: []string{sidecar1.InternalEndpoint("grpc"), sidecar2.InternalEndpoint("grpc")},
+		},
+		{
+			Endpoints: []string{receiver.InternalEndpoint("grpc")},
+			EndpointsSD: []httpconfig.FileSDConfig{
+				{
+					Files:           []string{fileSDPath},
+					RefreshInterval: model.Duration(time.Minute),
+				},
+			},
+		},
+		{
+			Endpoints: []string{sidecar3.InternalEndpoint("grpc"), sidecar4.InternalEndpoint("grpc")},
```

Comment:
Right, so I think these are duplicated stores. We already pass them in fileSD (although that isn't functional yet). But this is why this test can still pass. 🙂

Reply:
Yes! Realized it afterwards, removed the endpoints as we want them in fileSD